man/man1/fmt.1
.TH FMT 1
.SH NAME
fmt, htmlfmt \- simple text formatters
.SH SYNOPSIS
.B fmt
[
.I option ...
]
[
.I file ...
]
.PP
.B htmlfmt
[
.B -a
] [
.B -c
.I charset
] [
.B -I
] [
.B -u
.I url
] [
.I file ...
]
.SH DESCRIPTION
.I Fmt
copies the given
.I files
(standard input by default)
to its standard output, filling and indenting lines.
The options are
.TP
.BI -l " n
Output line length is
.IR n ,
including indent (default 70).
.TP
.BI -w " n
A synonym for
.BR -l .
.TP
.BI -i " n
Indent
.I n
spaces (default 0).
.TP
.BI -j
Do not join short lines: only fold long lines.
.PP
Empty lines and initial white space in input lines are preserved.
Empty lines are inserted between input files.
.PP
.I Fmt
is idempotent: it leaves already formatted text unchanged.
.PP
.I Htmlfmt
performs a similar service, but accepts as input text formatted with
HTML tags.
It accepts
.IR fmt 's
.B -l
and
.B -w
flags and also:
.TP
.BI -a
Normally
.I htmlfmt
suppresses the contents of form fields and anchors (URLs and image files); this flag
causes it to print them, in square brackets.
.TP
.BI -c " charset
change the default character set from iso-8859-1 to
.IR charset .
This is the character set assumed if there isn't one
specified by the html itself in a <meta> directive.
.TP
.B -I
When displaying anchors, suppress the printing of
image anchors.
.TP
.BI -u " url
Use
.I url
as the base URL for the document when displaying anchors; sets
.BI -a .
.SH SOURCE
.B \*9/src/cmd/fmt.c
.PP
.B \*9/src/cmd/htmlfmt
.SH BUGS
.I Htmlfmt
makes no attempt to render the two-dimensional geometry of tables;
it just treats the table entries as plain, to-be-formatted text.


CHANGES
November 26, 2005
	9term: add button 3 menu
	include: move utf.h, fmt.h from u.h to libc.h
	dist: print summary w/ errors during INSTALl

November 25, 2005
	root: add Makefile and configure for the blind,
		rename NOTES to README	

November 21, 2005
	lib9, 9pserve, vac: 9P2000.u fixes (Lucho Ionkov)

November 15, 2005
	libdraw: more X11 fixes
	libc.h: sched_yield NetBSD tweak (Anselm Garbe)

November 12, 2005
	faces: fix face location
	jpg, strings: remove references to /fd/

November 12, 2005
	various: X11 fixes

November 11, 2005
	9term: more fixes (Tim Wiess)

November 10, 2005
	9term: handle \r, other fixes (Tim Wiess)
	auxstats: new Darwin tweaks

November 9, 2005
	libthread: fix OS X context switch code

November 7, 2005
	vbackup: various fixes

November 6, 2005
	9term: new port directly from rio
	include/libc.h: p9syslog, not syslog in varargck
	libauthsrv: close fd, memory leak in authdial
	libdraw: X11 debugging helps
	lib9: add getcallerpc-arm (Ron Minnich)
	auxstats: Fix FreeBSD 6 build (William Josephson)

November 4, 2005
	stats: add -C to argchars (Lou Kamenov)
	libthread: ARM fixes (Ron Minnich), getcontext fixes

November 2, 2005
	vbackup: longer timeout on BSD mount
	libthread: Linux ARM fixes (Ron Minnich)

November 1, 2005
	various: ctype sign fixes (Tom Miller)
	libfmt: avoid overflow on %f
	libthread: Linux ARM fixes (Ron Minnich), FreeBSD 5 fixes
	libventi: various cleanup

October 31, 2005
	acme: preparation for multiline tags (Paul Lalonde)
	bc: 64-bit safe (Taj Khattra)
	faces: initial port (John Cummings)
	stats: add -F option (Lou Kamenov)
	various: ctype sign fixes (Tom Miller)

October 30, 2005
	auxstats: more OS X fixes (Jeff Sickel)	

October 29, 2005
	9p: clunk fids before exiting
	mk: handle MKSHELL type correctly
	upas, acme mail: add initial port (John Cummings)
	vbackup: silence compiler warnings
	venti: silence compiler warnings
	lib9: remove reference to ffork
		print correct file name in truerand error
	libdisk: add Disk routines
	libdiskfs: silence compiler warnings
	libframe: add noredraw flag
	libhtml: many fixes
	libsunrpc: silence compiler warnings

October 18, 2005
	9l: add autoframework (like autolib) for OS X (Jeff Sickel)
	page: use "9 file" always
	ps, psu: narrower display
	stats: better OS X support (Jeff Sickel)

October 17, 2005
	cb: undef old type macros before redefining.

October 5, 2005
	stats: better OS X support (Jeff Sickel) 

September 30, 2005
	libhtml: more &names and numbers (Erik Quanstrom)

September 29, 2005
	libc.h: add #pragmas for varargck, #ifdef'ed out.

September 26, 2005
	libthread: print thread info on SIGQUIT, SIGINFO, logging
	acme: add more thread names

September 21, 2005
	9term: fix SunOS typo (Bengt Kleberg, glaive)

September 19, 2005
	libhtml: more &names (Erik Quanstrom)

September 19, 2005
	lib9/truerand: OpenBSD lets you open /dev/random
		but not read it! (Chad Dougherty, Mechiel Lukkien)

September 15, 2005
	libdraw/x11-event: Ignore unexpected events (Rob Pike)
	jpg: do not revert to CMAP for second file (Rob Pike)

September 13, 2005
	9pserve: signedness bug

September 12, 2005
	dist: various attempts at Debian packaging
	lib/moveplan9.sh: allow specification of old directory
	Getdir: interpret tags with $ in them
	mpm: use g++ as LD, add tmac.pm to tmac directory
	include/libc.h, fcall.h: add 9P2000.u extensions
	9pserve: add 9P2000.u extensions
	vac: add 9P2000.u extensions, signedness bug

September 9, 2005
	lc: add -p like man page says
	various: insert (uchar) casts for ctype arguments
		for broken NetBSD.

September 4, 2005
	dist/main.html: note x86 OpenBSD

August 31, 2005
	file: spell IA-64 correctly (Zoltan Jarai)

August 30, 2005
	dist/debian, dist/pkg: start playing with building packages
	dist/isum: ignore 'loop not entered at top' (SunOS cc)
	dist/ss.html: add screen shots
	dist/unix.html: update VN license summary
	9l: no need to handle OpenBSD specially
	file: handle #! scripts, ELF better
	font/lucidasans/bold*unicode.7.font: more subfonts
	g: search .pl, .cgi
	hget: add hget(1)
	lp: add preliminary lp (not expected to work yet)
	src: use Plan 9 file(1)

August 12, 2005
	lib/moveplan9.files: add bin/stack
	include/libc.h: use DM constants from v9fs, add DMTMP
	libthread: add Linux-power-asm.s
	cb: add

August 11, 2005
	various: silence warnings
	unix/*: note new BSD license from Vita Nuova
	libdiskfs: add UFS2 support
	libthread: add NetBSD power assembly
	9term: send tty's current interrupt character on DEL keystroke
	rc: add limited rfork command, make sure errors propagate
	venti(8): document bloom line in config
	vftp: add get command
	vnfs: use strtoull for arguments instead of atoi.

August 10, 2005
	src/cmd/mkfile: do not build lp yet
	dist/addsrclinks, fixurls: use /usr/bin/env perl to find perl (NetBSD)
	lib9/_p9dir: avoid 32-bit overflow in old Linux size calculation
	lib9/truerand: use /dev/srandom on OpenBSD
	libthread: annoying wrappers for OpenBSD
	libmp: different assembly comments for old GNU as
	9p: set exit status
	file: add, appease gcc
	getflags: added
	mk: decide mtime-related race conservatively
	postscript/text2post: added
	rcmain: make fn cd safer for use with rc -e (in mk)
	rio: turn off debugging prints, remove 9wm.man, fixes from Andrey
	vbackup: NetBSD fixes for mount-BSD.c
	venti/srv: add <sys/stat.h> header for FreeBSD

July 28, 2005
	acme: subtle winunlock bug fix
	libdiskfs: fix double, triple indirect block reading for FFS
	vftp: new program
	vbackup: coalesce mount-*BSD files.

July 27, 2005
	include/u.h, src/libthread/thread.c: SunOS 10 attempt
	libdraw: less graceful, less raceful shutdown
	libthread: avoid C library races during threadexitsall
	awd: cannot use hostname -s on SunOS
	9pserve, srv: use correct type for NOFID

July 26, 2005
	various: fixes for SunOS from Bengt Kleberg

July 25, 2005
	vac: avoid infinite redirection while sniffing type of archive

July 24, 2005
	venti: many minor changes, updates

src/lib9/fmt/fmt.c
/* Copyright (c) 2002-2006 Lucent Technologies; see LICENSE */
#include <stdarg.h>
#include <string.h>

/*
 * As of 2020, older systems like RHEL 6 and AIX still do not have C11 atomics.
 * On those systems, make the code use volatile int accesses and hope for the best.
 * (Most uses of fmtinstall are not actually racing with calls to print that lookup
 * formats. The code used volatile here for years without too many problems,
 * even though that's technically racy. A mutex is not OK, because we want to
 * be able to call print from signal handlers.)
 *
 * RHEL is using an old GCC (atomics were added in GCC 4.8).
 * AIX is using its own IBM compiler (XL C).
 */
#if __IBMC__ || !__clang__ && __GNUC__ && (__GNUC__ < 4 || (__GNUC__==4 && __GNUC_MINOR__<8))
#warning not using C11 stdatomic on legacy system
#define _Atomic volatile
#define atomic_load(x) (*(x))
#define atomic_store(x, y) (*(x)=(y))
#define ATOMIC_VAR_INIT(x) (x)
#else
#include <stdatomic.h>
#endif

#include "plan9.h"
#include "fmt.h"
#include "fmtdef.h"

enum
{
	Maxfmt = 128
};

typedef struct Convfmt Convfmt;
struct Convfmt
{
	int	c;
	Fmts	fmt;
};

static struct
{
	/*
	 * lock updates to fmt by calling __fmtlock, __fmtunlock.
	 * reads can start at nfmt and work backward without
	 * further locking. later fmtinstalls take priority over earlier
	 * ones because of the backwards loop.
	 * once installed, a format is never overwritten.
	 */
	_Atomic int	nfmt;
	Convfmt	fmt[Maxfmt];
} fmtalloc = {
	#ifdef PLAN9PORT
		ATOMIC_VAR_INIT(27),
	#else
		ATOMIC_VAR_INIT(30),
	#endif
	{
		{' ',	__flagfmt},
		{'#',	__flagfmt},
		{'%',	__percentfmt},
		{'\'',	__flagfmt},
		{'+',	__flagfmt},
		{',',	__flagfmt},
		{'-',	__flagfmt},
		{'C',	__runefmt},	/* Plan 9 addition */
		{'E',	__efgfmt},
	#ifndef PLAN9PORT
		{'F',	__efgfmt},	/* ANSI only */
	#endif
		{'G',	__efgfmt},
	#ifndef PLAN9PORT
		{'L',	__flagfmt},	/* ANSI only */
	#endif
		{'S',	__runesfmt},	/* Plan 9 addition */
		{'X',	__ifmt},
		{'b',	__ifmt},		/* Plan 9 addition */
		{'c',	__charfmt},
		{'d',	__ifmt},
		{'e',	__efgfmt},
		{'f',	__efgfmt},
		{'g',	__efgfmt},
		{'h',	__flagfmt},
	#ifndef PLAN9PORT
		{'i',	__ifmt},		/* ANSI only */
	#endif
		{'l',	__flagfmt},
		{'n',	__countfmt},
		{'o',	__ifmt},
		{'p',	__ifmt},
		{'r',	__errfmt},
		{'s',	__strfmt},
	#ifdef PLAN9PORT
		{'u',	__flagfmt},
	#else
		{'u',	__ifmt},
	#endif
		{'x',	__ifmt},
	}
};

int	(*fmtdoquote)(int);

/*
 * __fmtlock() must be set
 */
static int
__fmtinstall(int c, Fmts f)
{
	Convfmt *p;
	int i;

	if(c<=0 || c>=65536)
		return -1;
	if(!f)
		f = __badfmt;

	i = atomic_load(&fmtalloc.nfmt);
	if(i == Maxfmt)
		return -1;
	p = &fmtalloc.fmt[i];
	p->c = c;
	p->fmt = f;
	atomic_store(&fmtalloc.nfmt, i+1);

	return 0;
}

int
fmtinstall(int c, int (*f)(Fmt*))
{
	int ret;

	__fmtlock();
	ret = __fmtinstall(c, f);
	__fmtunlock();
	return ret;
}

static Fmts
fmtfmt(int c)
{
	Convfmt *p, *ep;

	ep = &fmtalloc.fmt[atomic_load(&fmtalloc.nfmt)];
	for(p=ep; p-- > fmtalloc.fmt; )
		if(p->c == c)
			return p->fmt;

	return __badfmt;
}

void*
__fmtdispatch(Fmt *f, void *fmt, int isrunes)
{
	Rune rune, r;
	int i, n;

	f->flags = 0;
	f->width = f->prec = 0;

	for(;;){
		if(isrunes){
			r = *(Rune*)fmt;
			fmt = (Rune*)fmt + 1;
		}else{
			fmt = (char*)fmt + chartorune(&rune, (char*)fmt);
			r = rune;
		}
		f->r = r;
		switch(r){
		case '\0':
			return nil;
		case '.':
			f->flags |= FmtWidth|FmtPrec;
			continue;
		case '0':
			if(!(f->flags & FmtWidth)){
				f->flags |= FmtZero;
				continue;
			}
			/* fall through */
		case '1': case '2': case '3': case '4':
		case '5': case '6': case '7': case '8': case '9':
			i = 0;
			while(r >= '0' && r <= '9'){
				i = i * 10 + r - '0';
				if(isrunes){
					r = *(Rune*)fmt;
					fmt = (Rune*)fmt + 1;
				}else{
					r = *(char*)fmt;
					fmt = (char*)fmt + 1;
				}
			}
			if(isrunes)
				fmt = (Rune*)fmt - 1;
			else
				fmt = (char*)fmt - 1;
		numflag:
			if(f->flags & FmtWidth){
				f->flags |= FmtPrec;
				f->prec = i;
			}else{
				f->flags |= FmtWidth;
				f->width = i;
			}
			continue;
		case '*':
			i = va_arg(f->args, int);
			if(i < 0){
				/*
				 * negative precision =>
				 * ignore the precision.
				 */
				if(f->flags & FmtPrec){
					f->flags &= ~FmtPrec;
					f->prec = 0;
					continue;
				}
				i = -i;
				f->flags |= FmtLeft;
			}
			goto numflag;
		}
		n = (*fmtfmt(r))(f);
		if(n < 0)
			return nil;
		if(n == 0)
			return fmt;
	}
}
